### PR TITLE
[sdk] Nullable annotations for the metrics folder

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Added nullable annotations for things missing in the Metrics folder of the OpenTelemetry project.
* Enable nullable reference types globally for the OpenTelemetry project.

Note: I'll do a follow-up to remove `#nullable enable` from all the files.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)

### CODE REVIEW

File: src/OpenTelemetry/OpenTelemetry.csproj [hunk 1]
```diff
@@ -3,9 +3,6 @@
     <TargetFrameworks>$(DefaultTargetFrameworks);netstandard2.1</TargetFrameworks>
     <Description>OpenTelemetry .NET SDK</Description>
     <MinVerTagPrefix>core-</MinVerTagPrefix>
-
-    <!-- this is temporary. will remove in future PR. -->
-    <Nullable>disable</Nullable>
```
Comment (CodeBlanch) [inline src/OpenTelemetry/OpenTelemetry.csproj:8]: 🥂 🚀 🥳

**Comment by CodeBlanch** [comment]
Tag #3958

File: src/OpenTelemetry/OpenTelemetry.csproj (hunk 1 above)
Comment (Kielek) [reply src/OpenTelemetry/OpenTelemetry.csproj:8]: Please check if you can remove `#nullable enable` from all files in this projec (the alternative approach is to do it, when all projects in repository will have nullable enabled).

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs [hunk 2]
```diff
@@ -30,13 +32,16 @@ internal List<Metric> AddMetricsWithNoViews(Instrument instrument)
         for (var cur = this.Head; cur != null; cur = cur.Next)
         {
             var metric = cur.Value.AddMetricWithNoViews(instrument);
-            metrics.Add(metric);
+            if (metric != null)
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:35]: Why this change? Does it not affect the users of this list?

File: src/OpenTelemetry/Metrics/MetricPoint.cs [hunk 3]
```diff
@@ -252,9 +252,14 @@ internal MetricPointStatus MetricPointStatus
             this.ThrowNotSupportedMetricTypeException(nameof(this.GetHistogramSum));
         }
 
+        Debug.Assert(
+            this.mpComponents?.HistogramBuckets != null
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MetricPoint.cs:256]: Is this check needed? The return statement that follows this checks it anyway.

File: src/OpenTelemetry/Metrics/MetricPoint.cs [hunk 4]
```diff
@@ -294,7 +301,9 @@ public ExponentialHistogramData GetExponentialHistogramData()
             this.ThrowNotSupportedMetricTypeException(nameof(this.GetExponentialHistogramData));
         }
 
-        return this.mpComponents!.Base2ExponentialBucketHistogram.GetExponentialHistogramData();
+        Debug.Assert(this.mpComponents?.Base2ExponentialBucketHistogram != null, "Base2ExponentialBucketHistogram was null");
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MetricPoint.cs:304]: Use `this.mpComponents!` in `Debug.Assert`?

File: src/OpenTelemetry/Metrics/MetricPoint.cs [hunk 5]
```diff
@@ -844,7 +865,9 @@ internal void TakeSnapshot(bool outputDelta)
                 {
                     var histogramBuckets = this.mpComponents!.HistogramBuckets;
 
-                    AcquireLock(ref histogramBuckets.IsCriticalSectionOccupied);
+                    Debug.Assert(this.mpComponents.HistogramBuckets != null, "HistogramBuckets was null");
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MetricPoint.cs:868]: Could we simplify this?
```suggestion
                    Debug.Assert(histogramBuckets != null, "HistogramBuckets was null");
```

File: src/OpenTelemetry/OpenTelemetry.csproj (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/OpenTelemetry.csproj:8]: I'm going to do that as a follow-up so I don't explode the diff. Originally I wasn't going to enable at the project-level as part of this PR but I decided that is the most important part in order to stop the bleeding of code being added without annotations 😄

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:35]: I think it is a (rare) bug fix. This code is returning a list of all metrics published by the inner readers. Before it just assumed that each inner reader would return someting, but the inner reader may return `null` from `AddMetricWithNoViews`:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/c97a233a9a9d9656b8c431f69f5697f90554b3bb/src/OpenTelemetry/Metrics/MetricReaderExt.cs#L68

https://github.com/open-telemetry/opentelemetry-dotnet/blob/c97a233a9a9d9656b8c431f69f5697f90554b3bb/src/OpenTelemetry/Metrics/MetricReaderExt.cs#L84

Comment (utpilla) [reply src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs:35]: I think this accounted for in the method that consumes this list. If it's not required for enabling `nullable`, the I would strongly suggest we remove this change from this PR. There are already a lot changes in this PR. If this is indeed a bug that we need to fix, let's create a dedicated PR for it with the required unit tests added.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/c97a233a9a9d9656b8c431f69f5697f90554b3bb/src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs#L46-L49

File: src/OpenTelemetry/Metrics/StringArrayEqualityComparer.cs [hunk 6]
```diff
@@ -50,8 +54,11 @@ public bool Equals(string[] strings1, string[] strings2)
 
     public int GetHashCode(string[] strings)
     {
+        Debug.Assert(strings != null, "strings was null");
+
 #if NET6_0_OR_GREATER
         HashCode hashCode = default;
+
         for (int i = 0; i < strings.Length; i++)
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/StringArrayEqualityComparer.cs:62]: Use `strings!.Length` here as well?

File: src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/CompositeMetricReaderExt.cs]: Interesting! OK I updated it so `null` is valid in the list. We should probably switch this to drop the `null`s and not add them to the list so that during measurement we don't need to waste cycles checking invalid indexes but fine to do as a follow-up.

File: src/OpenTelemetry/Metrics/StringArrayEqualityComparer.cs (hunk 6 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/StringArrayEqualityComparer.cs:62]: We could, but it isn't needed. A lot of the `!`s you see are just to make net462 happy. .NET6+ targets have better decorations across the runtime so the analysis gets smarter. In this case, we call...

```csharp
public static void Assert([DoesNotReturnIf(false)] bool condition, string? message)
```

...and the `DoesNotReturnIf` for net6.0 is what clears the warning.

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricPoint.cs:256]: In the return we do a couple dereference ops without checks:

* `this.mpComponents!`
* `this.mpComponents.Base2ExponentialBucketHistogram!`

This assert is to make sure we have a valid path through what the return might do.

We could rewrite it like this:

```csharp
        Debug.Assert(this.mpComponents != null, "this.mpComponents was null");

        if (this.mpComponents!.HistogramBuckets != null)
        {
            return this.mpComponents!.HistogramBuckets.SnapshotSum;
        }

        Debug.Assert(this.mpComponents.Base2ExponentialBucketHistogram != null, "this.mpComponents.Base2ExponentialBucketHistogram was null");

        return this.mpComponents.Base2ExponentialBucketHistogram!.SnapshotSum;
```

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricPoint.cs:304]: My goal here is to assert in debug all the stuff we just assume to be non-null and access with suppression using `!`.

So I wouldn't do this...

```csharp
Debug.Assert(this.mpComponents!.Base2ExponentialBucketHistogram != null, "Base2ExponentialBucketHistogram was null");
```

Because if someone made a mistake upstream and `this.mpComponents` somehow ended up `null` it could yield a NRE.

But we could do this...

```csharp
        Debug.Assert(this.mpComponents != null, "this.mpComponents was null");
        Debug.Assert(this.mpComponents!.HistogramBuckets != null, "this.mpComponents.HistogramBuckets was null");
```

The reason I went with the `?` in there is it more or less does the work of the two assertions in one shot.

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricPoint.cs]: I just went through and tried to simplify as many as I could.

**Review by utpilla — APPROVED** [review]
